src/onnxruntime.cc

```diff
@@ -682,10 +682,40 @@ ModelState::AutoCompleteConfig()
   OrtAllocator* default_allocator;
   std::string model_path;
   {
+    TRITONSERVER_InstanceGroupKind kind;
+
+    triton::common::TritonJson::Value instance_group;
+    ModelConfig().Find("instance_group", &instance_group);
+
+    // Earlier in the model lifecycle, device checks for the instance group
+    // have already occurred. If at least one instance group with
+    // "kind" = "KIND_GPU" then allow model to use GPU else autocomplete to CPU
+    bool found_gpu_instance = false;
+    for (size_t i = 0; i < instance_group.ArraySize(); ++i) {
+      triton::common::TritonJson::Value instance_obj;
+      instance_group.IndexAsObject(i, &instance_obj);
+
+      triton::common::TritonJson::Value instance_group_kind;
+      instance_obj.Find("kind", &instance_group_kind);
+      std::string kind_str;
+      RETURN_IF_ERROR(instance_group_kind.AsString(&kind_str));
+
+      if (kind_str == "KIND_GPU" || kind_str == "KIND_AUTO") {
+        found_gpu_instance = true;
+        break;
+      }
+    }
+
+    if (found_gpu_instance) {
+      kind = TRITONSERVER_INSTANCEGROUPKIND_GPU;
+    } else {
```

Comment:
1. Isn't the problem in https://github.com/triton-inference-server/server/issues/4346#issuecomment-1119885043 because default KIND is KIND_AUTO ? Wouldn't this code still default to GPU if no kind is specified even on a CPU machine? Or can you point me to the code that sets KIND to KIND_CPU if no GPUs are available? (or include a comment detailing similarly)

2. Should this check be done in `LoadModel()` instead since it can be passed KIND_AUTO even if GPU isn't necessarily supported? Currently KIND_AUTO == KIND_GPU in LoadModel(), and it seems more like KIND_AUTO should specify that we pick GPU if available, otherwise pick CPU.

Reply:
1. The previous code I was referring to was [here]((https://github.com/triton-inference-server/core/blob/2ba05e79e8f5db7c354f2ffb48cc8cd920c23be7/src/model_config_utils.cc#L749). If the model config gets loaded with KIND_AUTO, then the linked code checks for devices which it found and assigns the instance group kind. In the case of the bug this is solving, the `supported_gpus` set will be empty since no devices were found and the group instance kind will be set to KIND_CPU.  "KIND_AUTO" should never make it to the backend, which should address your concern.
2. I hesitate putting this into `LoadModel()` since `ModelInstanceState()` also [calls this function](https://github.com/triton-inference-server/onnxruntime_backend/blob/493d43ef21906df32ec80d99676f19adfc2176af/src/onnxruntime.cc#L949), and I'm not certain what code is getting executed before this. Also, both the offline channel discussion and the linked github issue both specify using `--strict-model-config=false` which is what this PR aims to solve. I can look further into this if you wish.

Reply:
re: (2), Keeping the logic in Autocomplete is OK for now

Reply:
re: (1) in your testing, can you confirm that loading a model config that **doesn't** specify instance_group with `--strict-model-config=true` will correctly use CPU if no GPU is available?

You should be able to use our example [densenet_onnx](https://github.com/triton-inference-server/server/blob/main/docs/examples/model_repository/densenet_onnx/config.pbtxt) model and config if you don't have any others handy.

Reply:
Will check this out

Reply:
Tested this out with the "resnet50" model (not the cpu verison) and used the following command: `$tritonserver --model-repository=/path/to/models --strict-model-config=true --log-verbose=1` and following occurred:
1. Model was successfully loaded 
2. When queried, the model configuration has the following:
  a.   ```  "instance_group": [{
            "name": "resnet50",
            "kind": "KIND_CPU",
            "count": 1,
            "gpus": [],
            "secondary_devices": [],
            "profile": [],
            "passive": false,
            "host_policy": ""
        }
    ],```

When adding the following to the config file:
```
instance_group [
    {
      count: 1
      kind: KIND_GPU
      gpus: [ 0 ]
    }
]
```
and then executing `$tritonserver --model-repository=/path/to/models --strict-model-config=true --log-verbose=1` I received the following:
1. The model did not load
2. Triton returned the expected error: `resnet50 specifies invalid or unsupported gpu id 0. GPUs with at least the minimum required CUDA compute compatibility of 6.000000`

Reply:
Awesome, thanks! I assume both an empty list or omitting the `gpus` field here would have similar behavior?

Reply:
1. Omitting the GPUs field with `--strict-model-config=true` constructed the correct model config with kind=KIND_CPU and model was loaded successfully.
2. Empty list had the same effect as (1)